cmd/cellrangerBam/cellrangerBam.go

```diff
@@ -0,0 +1,823 @@
+	var columns, columns2 []string
+	var j, join string
+	var n int
+
+	inChan, head := sam.GoReadToChan(a)
+	if *iteration == 1 {
+		bw = sam.NewBamWriter(out, head)
+	}
+	for i := range inChan {
+		err = sam.ParseExtra(&i)
+		columns = strings.Split(i.Extra, "\t")
+		for n, j = range columns {
+			columns2 = strings.Split(j, ":")
+			if columns2[0] == "CB" {
+				columns[n] = fmt.Sprintf("%s_%d", j, *iteration)
+				join = strings.Join(columns, "\t")
+				i.Extra = join
+			}
+		}
+		sam.WriteToBamFileHandle(bw, i, 0)
+	}
+	exception.PanicOnErr(err)
+	*iteration++
+	return bw
+}
+
+// binnedPseudobulk is the psuedobulk function if the -binCells option is used. It adds an addition column to the dataframe corresponding to bin identity
+func binnedPseudobulk(inSlices [][]string, out *fileio.EasyWriter, norm string) {
+	var j, i string
+	var count float64
+	var found bool
+	var columns []string
+
+	whichBin := 'A'
+	fileio.WriteToFileHandle(out, "construct\tcounts\tbin")
+	for _, bin := range inSlices {
+		mp := make(map[string]float64)
+		var toWrite []string
+		for _, j = range bin {
+			columns = strings.Split(j, "\t")
+			count, found = mp[columns[1]]
+			if !found {
+				mp[columns[1]] = 1
+			} else {
+				mp[columns[1]] = count + 1
+			}
+		}
+		if norm != "" {
+			inputNormalize(mp, norm)
+		}
+		for i = range mp {
+			toWrite = append(toWrite, fmt.Sprintf("%s\t%f\t%c", i, mp[i], whichBin))
+		}
+		sort.Strings(toWrite)
+		for _, i = range toWrite {
+			fileio.WriteToFileHandle(out, i)
+		}
+		whichBin++
+	}
+}
+
+// determine ideal bins will determine how many bins should be used for the pseudobulk -binCells option
+func determineIdealBins(s Settings, cellTypeSlice []string) int {
+	var count, j int
+	var found, stop bool
+	var z, l, m string
+	var binSlice []int
+	var matrix [][]string
+
+	nc := fileio.Read(s.determineBins)
+
+	for i := 0; i <= 10; i++ { //do this whole loop 10 times (essentially 10 replicates)
+		stop = false
+		for j = 1; j <= 100; j++ { //loop from bins 1 to 100 (j)
+			s.binCells = j                                   //set the binCells parameter to j
+			matrix = distributeCells(s, cellTypeSlice, true) //run distributeCells with that value of j
+			for k := range matrix {                          //loop through bins in matrix
+				ncMap := make(map[string]int) //create a map that has ncConstructName--counts
+				for _, z = range nc {         //populate the map
+					ncMap[z] = 0
+				}
+				for _, l = range matrix[k] { //loop through the constructs in a particular bin
+					columns := strings.Split(l, "\t") //split an entry in from the matrix
+					count, found = ncMap[columns[1]]  //search the 2nd column to see if it is a negative control
+					if found {                        //if it is, add a count to the map for that particular negative control
+						ncMap[columns[1]] = count + 1
+					}
+				}
+				for m = range ncMap {
+					count, _ = ncMap[m]
+					if count < 1 {
+						binSlice = append(binSlice, j-1)
```

Comment:
How is the ideal bin number determined? It looks like ideal bin number = average of binSlice []int, and binSlice records the j (number of bins, some number between 1 and 100) that makes ncMap[m] count <1 (gets 0 counts for negative control constructs) in 10 trials?

Reply:
yes that is correct